enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,515 @@
+deployed. As part of the start procedure, MicroShift will configure the router
+with fixed parameters, such as being always enabled, and tied to ports 80 and
+443. None of these are configurable for the user, and they also tie to how the
+router is exposed in the cluster.
+
+In order to allow more flexibility, the following configuration parameters shall
+be added:
+- Enable/Disable. Some use cases may be egress only, meaning they do not need
+  an ingress controller. In this case the router pod should not be scheduled
+  to save resources. There should be no rules associated to firewalld or
+  iptables either for an improved security posture.
+- Listening ports. Allow configuring ports other than 80 and 443 improves
+  flexibility.
+- Which IPs the router listens on. Some use cases may require the router to be
+  reachable only from certain networks.
+
+### User Stories
+As a MicroShift user, I want to enable the default router with a configuration
+option.
+
+As a MicroShift user, I want to disable the default router and have all the
+manifests automatically removed.
+
+As a MicroShift user, I want to be able to configure the ports on which the
+default router is listening.
+
+As a MicroShift user, I want to be able to choose the IPs where the router is
+listening.
+
+As a MicroShift user, I want to be able to allow traffic to the router from
+specific IP addresses.
+
+### Goals
+* Allow users to enable the router.
+* Allow users to disable the router.
+* Allow users to reach the router in the configured ports.
+* Allow users to configure in which IPs the router listens.
+* Allow users to allow traffic from specific IP addresses.
+* Allow users to deny traffic to the router.
+* Internal access from applications to the router must remain unchanged.
+
+### Non-Goals
+N/A
+
+## Proposal
+Each configuration option will get its own section with specifics and
+details.
+As a common baseline, a new top level section will be added to the
+configuration:
+```yaml
+apiServer:
+   ...
+# new from here.
+ingress:
+  status: <Enabled|Disabled> # Defaults to Enabled.
+  ports:
+    http: <int> # Defaults to 80.
+    https: <int> # Defaults to 443.
+  expose: # Defaults to IPs in the host. Details below.
+    hostnames:
+    - <hostname>
+    interfaces:
+    - <NIC name>
+    ipAddresses:
+    - <IP address>
+```
+
+### Enable/disable the router
+The following configuration is proposed:
+```yaml
+ingress:
+  status: <Enabled|Disabled> # Defaults to Enabled.
+```
+
+With this option MicroShift will decide whether it should create the router
+upon starting. This includes not just the pod, but all the associated resources
+that come along with it: namespace, services, configmaps, etc.
+
+When setting the option to `Disabled`, the next MicroShift restart will
+delete all the default router related resources.
+
+Note that any existing routes will stop being served and it will require an
+application deploying an OpenShift router instance.
```

Comment:
Hm, this is confusing.  Are you saying that if you disable the default router but still want to use routes, then you (most likely meaning the cluster admin) need to deploy a custom router instance?  I think the phrasing "it will require an application deploying an OpenShift router instance" is throwing me; it suggests to me that the application itself is taking the action of deploying the router.

Reply:
We would be more likely to encourage someone to deploy a different ingress controller, right? The router stuff is an implementation detail of our ingress, even though it's technically exposed in its own right.

But let's not get hung up on that. If the user disables the router, they won't have either ingress or the router. If that's the intent they express, we can assume they have a reason for it, like their workload not needing it.